File: 16-3sum-closest/3sum-closest.go


func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	n := len(nums)

	closestSum := nums[0] + nums[1] + nums[2]

	for i := 0; i < n-2; i++ {
		left := i + 1
		right := n - 1

		for left < right {
			currSum := nums[i] + nums[left] + nums[right]

			if abs(target-currSum) < abs(target-closestSum) {
				closestSum = currSum
			}

			if currSum < target {
				left++
			} else if currSum > target {
				right--
			} else {
				return target 
			}
		}
	}

	return closestSum
}

func abs(x int) int {
	return int(math.Abs(float64(x)))
}

File: 16-3sum-closest/3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int diff = 1e5;
        int ans = 0;
        for (int k = 0; k < nums.size(); k++) {
            for (int i = k + 1; i < nums.size(); i++) {
                for (int j = i + 1; j < nums.size(); j++) {
                    if (abs(target - (nums[i] + nums[j] + nums[k])) < diff) {
                        diff = abs(target - (nums[i] + nums[j] + nums[k]));
                        ans = nums[i] + nums[j] + nums[k];
                    }
                }
            }
        }
        return ans;
    }
};